Use WeakPtr to store CSSFontFace::Client and FontSelector

rniwa opened the pull request:
#### 0b2ad9a154b0e9e3b86c932f976164c2cbbfc90f
<pre>
Use WeakPtr to store CSSFontFace::Client and FontSelector
<a href="https://bugs.webkit.org/show_bug.cgi?id=250756">https://bugs.webkit.org/show_bug.cgi?id=250756</a>

Reviewed by Chris Dumez.

Use WeakPtr instead of raw pointers to keep track of CSSFontFace::Client and FontSelector.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::iterateClients):
(WebCore::CSSFontFace::addClient):
(WebCore::CSSFontFace::removeClient):
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/FontFace.h:
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::addClient):
(WebCore::FontCache::removeClient):
(WebCore::FontCache::invalidate):
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontSelector.h:

Canonical link: <a href="https://commits.webkit.org/259118@main">https://commits.webkit.org/259118@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8b5e8edc2fa889caf1e8bfe8052237dac6bf7110

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36683 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112965 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/173286 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/107690 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13877 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3750 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95992 "Built successfully") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112099 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10705 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93764 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38418 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92529 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25370 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80072 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6215 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26766 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6389 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3288 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46285 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8150 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3328 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/css/CSSFontFace.cpp:
```diff
@@ -52,10 +52,15 @@ namespace WebCore {
 {
-    for (auto& client : copyToVectorOf<RefPtr<CSSFontFace::Client>>(clients))
-        callback(*client);
+    Vector<Ref<CSSFontFace::Client>> clientsCopy;
```
Any reason you chose not to use WeakHashSet::forEach() which does the copy for you? Is it because you really want to Ref the clients?

cdumez replied:
If you really want to Ref, maybe `copyToVectorOf<Ref<CSSFontFace::Client>>(clients)` would work and would be more concise.

rniwa replied:
`copyToVectorOf` doesn't work because it `WeakHashSet` doesn't have `size()`.

rniwa replied:
I guess I can add a template specialization for it.

rniwa replied:
Oh... actually, copyToVectorOf works out of box. I just had to change the type to Ref<CSSFontFace::Client> instead of RefPtr<CSSFontFace::Client>.

cdumez commented on Source/WebCore/platform/graphics/FontCache.cpp:
```diff
@@ -397,7 +396,11 @@ void FontCache::invalidate()
     ++m_generation;
 
-    for (auto& client : copyToVectorOf<RefPtr<FontSelector>>(m_clients))
+    Vector<Ref<FontSelector>> clientList;
```
Ditto.

rniwa committed:
Use WeakPtr to store CSSFontFace::Client and FontSelector
https://bugs.webkit.org/show_bug.cgi?id=250756

Reviewed by Chris Dumez.

Use WeakPtr instead of raw pointers to keep track of CSSFontFace::Client and FontSelector.

* Source/WebCore/css/CSSFontFace.cpp:
(WebCore::iterateClients):
(WebCore::CSSFontFace::addClient):
(WebCore::CSSFontFace::removeClient):
* Source/WebCore/css/CSSFontFace.h:
* Source/WebCore/css/CSSFontSelector.h:
* Source/WebCore/css/FontFace.h:
* Source/WebCore/platform/graphics/FontCache.cpp:
(WebCore::FontCache::addClient):
(WebCore::FontCache::removeClient):
(WebCore::FontCache::invalidate):
* Source/WebCore/platform/graphics/FontCache.h:
* Source/WebCore/platform/graphics/FontSelector.h:

Canonical link: https://commits.webkit.org/259118@main
